### PR TITLE
Refactor temporal arithmetic 

### LINKED ISSUE #6155: arithmetic operations on timestamps fail when `now()` is used as the LHS argument
### Describe the bug

When performing an arithmetic operation on timestamps, `now()` causes an error when it is used as the LHS argument. It works as expected when it is used at the RHS argument.

```
Internal("If RHS of the operation is an array, then LHS also must be")
```

### To Reproduce

The following will return an error and then panic on unwrap.

```rust
let ctx = SessionContext::new();

let df = ctx
    .sql(
        r#"
WITH cte AS (
    SELECT now() AS dt 
)
SELECT now() - cte.dt FROM cte
"#,
    )
    .await
    .unwrap();
```

### Expected behavior

The operation should be successfully evaluated.

### Additional context

The following will run without issue. Note that `now()` is the RHS arg in this sample.

```rust
let ctx = SessionContext::new();

let df = ctx
    .sql(
        r#"
WITH cte AS (
    SELECT now() AS dt 
)
SELECT cte.dt - now() FROM cte
"#,
    )
    .await
    .unwrap();
```

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #6155.

# Rationale for this change

https://github.com/apache/arrow-datafusion/pull/6196 PR has solved the issue of temporal scalars cannot appear as the LHS argument in an arithmetic operation. But there is an unoptimized solution such as creating an array from a scalar value as the length of the RHS array, and apply array vs array operations. This PR solves this issue in a similar way of LHS:Array and RHS:Scalar arithmetic.

# What changes are included in this PR?

In the datetime.rs, there are two functions that handle arithmetic operations between arrays and arrays, and between scalars and arrays. To determine the appropriate operation to perform, a flag is used to hold information about the left-hand side (LHS) and right-hand side (RHS) operands.

In the kernel_arrow.rs, type matching is performed to identify the specific operation to execute. Finally, the actual computation is carried out in the corresponding function within the binary.rs.

# Are these changes tested?

Yes.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -1689,7 +1874,524 @@ pub fn interval_scalar_interval_op(
...
+}
+
+/// This function handles interval +/- interval operations where the former is
+/// a scalar and the latter is an array, resulting in an interval array.
+pub fn scalar_interval_op_interval(
+    scalar: &ScalarValue,
+    sign: i32,
+    array: &ArrayRef,
+) -> Result<ArrayRef> {
+    let ret = match (scalar, array.data_type()) {
```
Comment (jackwener) [inline datafusion/physical-expr/src/expressions/binary.rs:1887]: same with above.
We can use use ScalarValue::* use DataType::* use IntervalUnit::* to simplify code

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -1689,7 +1874,524 @@ pub fn interval_scalar_interval_op(
...
+        // YearMonth op YearMonth
+        (
+            ScalarValue::IntervalYearMonth(Some(lhs)),
+            DataType::Interval(IntervalUnit::YearMonth),
+        ) => {
+            let array = as_interval_ym_array(&array)?;
+            let ret: PrimitiveArray<IntervalYearMonthType> =
+                unary(array, |rhs| op_ym(*lhs, rhs, sign));
+            Arc::new(ret) as _
```
Comment (jackwener) [inline datafusion/physical-expr/src/expressions/binary.rs:1896]: `Arc::new(ret) as _` occurs many times, is there any way we can avoid it?

Comment (berkaysynnada) [reply datafusion/physical-expr/src/expressions/binary.rs:1834]: Thank you for the review. I cannot figure out getting rid of `as _ `, but your other advice of adding namespaces has significantly decreased the number of lines.

**Comment by alamb** [comment]
I plan to review this PR tomorrow. Perhaps @tustvold  is interested as I believe he is planning some work to improve the temporal kernel support in arrow-rs

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -1247,14 +1247,14 @@ pub fn calculate_naives<const TIME_MODE: bool>(
 }
 
 #[inline]
-pub fn date32_add(days: i32, scalar: &ScalarValue, sign: i32) -> Result<i32> {
+pub fn date32_op(days: i32, scalar: &ScalarValue, sign: i32) -> Result<i32> {
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1250]: 👍  that is a nice change

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 4]
```diff
@@ -333,34 +337,88 @@ pub(crate) fn add_dyn_temporal(left: &ArrayRef, right: &ArrayRef) -> Result<Arra
     }
 }
 
-pub(crate) fn add_dyn_temporal_scalar(
+pub(crate) fn add_dyn_temporal_right_scalar(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:340]: Since adding `interval + duration` is the same as adding `duration + interval` I wonder if we need both of these kernels? Instead, perhaps we can use the same pattern as used elsewhere in binary.rs and do something like:

Given: `a + b`:
* if `b` is a scalar, call `add_dyn_scalar(a, b)`
* if `a` is a scalar, call `add_dyn_scalar(b, a)`

Given: `a - b`:
* if `b` is a scalar, call `sub_dyn_scalar(a, b)`
* if `a` is a scalar, call `add_dyn_scalar(-b, a)`

🤔

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 5]
```diff
@@ -1087,84 +1092,69 @@ impl BinaryExpr {
         scalar: ScalarValue,
         result_type: &DataType,
     ) -> Result<Option<Result<ArrayRef>>> {
+        use Operator::*;
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:1095]: ❤️

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 6]
```diff
@@ -1689,7 +1812,500 @@ pub fn interval_scalar_interval_op(
             scalar.get_datatype(),
         )))?,
     };
-    Ok(ColumnarValue::Array(ret))
+    Ok(ret)
+}
+
+/// This function handles interval +/- interval operations where the former is
+/// a scalar and the latter is an array, resulting in an interval array.
+pub fn scalar_interval_op_interval(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:1820]: cc @tustvold  I am not sure if these exist (yet) in arrow-rs but I think we should be working to port them upstream

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 7]
```diff
@@ -1519,18 +1599,85 @@ pub fn ts_scalar_interval_op(
             array.data_type()
         )))?,
     };
-    Ok(ColumnarValue::Array(ret))
+    Ok(ret)
+}
+
+/// This function handles timestamp +/- interval operations where the former is
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:1605]: Minor suggestion is to move these functions into datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs so they are more discoverable

**Review by alamb — COMMENTED** [review]
Thank you @berkaysynnada  -- in general this code is looking quite good. I think it is structured nicely and is improving 

My one big concern is that the new code seems to be mostly be untested -- specifically there are a lot of new functions and code that looks similar to me such as `ts_ms_interval_mdn_op` and `ts_us_interval_mdn_op`

Do you know if  this code covered by existing tests? 

I realize it might be repetitive to write tests for this code, but I think it is is really important to avoid regressions in the future. Perhaps there is some way to reduce the code replication (and therefore also the testing burden) 🤔

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs (hunk 4 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:340]: A question that comes to my mind about this approach is whether the unary minus on the LHS (the array) would result in unnecessary computation or not. Given the complexity of the types involved, I am not sure if the compiler would be smart enough to optimize things all the way and emit simple subtraction code (vs. negating the entire array and then summing again).

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 6 above)
Comment (tustvold) [reply datafusion/physical-expr/src/expressions/binary.rs:1820]: I intend to spend some time designing what this might look like next week

**Comment by berkaysynnada** [comment]
> Thank you @berkaysynnada -- in general this code is looking quite good. I think it is structured nicely and is improving
> 
> My one big concern is that the new code seems to be mostly be untested -- specifically there are a lot of new functions and code that looks similar to me such as `ts_ms_interval_mdn_op` and `ts_us_interval_mdn_op`
> 
> Do you know if this code covered by existing tests?
> 
> I realize it might be repetitive to write tests for this code, but I think it is is really important to avoid regressions in the future. Perhaps there is some way to reduce the code replication (and therefore also the testing burden) 🤔

Thank you for the review ☺️ I have applied your suggestions. I have added some tests that compares the results of a(ScalarValue) vs b(another ScalarValue) and a vs c(Array having one element and the same value of b). I thought this would suffice, as the ScalarValue tests are wide-ranging and testing edge cases. I have also tried to reduce arithmetic functions as much as possible, but I would appreciate it if you let me know if anything else catches your eye.

**Comment by berkaysynnada** [comment]
> > Thank you @berkaysynnada -- in general this code is looking quite good. I think it is structured nicely and is improving
> > My one big concern is that the new code seems to be mostly be untested -- specifically there are a lot of new functions and code that looks similar to me such as `ts_ms_interval_mdn_op` and `ts_us_interval_mdn_op`
> > Do you know if this code covered by existing tests?
> > I realize it might be repetitive to write tests for this code, but I think it is is really important to avoid regressions in the future. Perhaps there is some way to reduce the code replication (and therefore also the testing burden) 🤔
> 
> Thank you for the review ☺️ I have applied your suggestions. I have added some tests that compare the results of a(ScalarValue) vs b(another ScalarValue) and a vs c(Array having one element and the same value of b). I thought this would suffice, as the ScalarValue tests are wide-ranging and testing edge cases. I have also tried to reduce arithmetic functions as much as possible, but I would appreciate it if you let me know if anything else catches your eye.

**Comment by alamb** [comment]
Thanks everyone!
